// @egjs/svelte-conveyer/dist/conveyer.esm.js
/*
Copyright (c) NAVER Corp.
name: @egjs/svelte-conveyer
license: MIT
author: NAVER Corp.
repository: https://github.com/naver/egjs-conveyer
version: 1.4.8
*/
import { REACTIVE_CONVEYER } from '@egjs/conveyer';
import { adaptReactive, camelize } from '@cfcs/core';
import { onMount, onDestroy } from 'svelte/internal';
import { writable } from 'svelte/store';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
var __assign = function () {
  __assign = Object.assign || function __assign(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
      s = arguments[i];

      for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
    }

    return t;
  };

  return __assign.apply(this, arguments);
};

/*
Copyright (c) NAVER Corp.
name: @cfcs/svelte
license: MIT
author: NAVER Corp.
repository: https://github.com/naver/cfcs
version: 0.0.13
*/

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
var __assign$1 = function () {
  __assign$1 = Object.assign || function __assign(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
      s = arguments[i];

      for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
    }

    return t;
  };

  return __assign$1.apply(this, arguments);
};

function useRef(defaultValue) {
  var refFunction = function (value) {
    refFunction.current = value;
    return {
      destroy: function () {
        return;
      }
    };
  };

  refFunction.current = defaultValue;
  return refFunction;
}
function useReactive(reactiveProps) {
  var adaptResult = adaptReactive(reactiveProps);
  var reactiveState = adaptResult.state();
  var writables = {};
  var names = Object.keys(reactiveState);
  var methods = adaptResult.methods();

  for (var name in reactiveState) {
    writables[name] = writable(reactiveState[name]);
  }

  onMount(function () {
    adaptResult.mounted();
    var inst = adaptResult.instance();
    names.forEach(function (name) {
      inst.subscribe(name, function (value) {
        writables[name].set(value);
      });
    });
    adaptResult.init();
  });
  onDestroy(function () {
    adaptResult.destroy();
  });
  var reactiveEvents = reactiveProps.events || [];
  var events = reactiveEvents.reduce(function (result, name) {
    result[camelize("on ".concat(name))] = function (callback) {
      onMount(function () {
        adaptResult.on(name, callback);
      });
      onDestroy(function () {
        adaptResult.off(name, callback);
      });
    };

    return result;
  }, {});
  return __assign$1(__assign$1(__assign$1({}, writables), methods), events);
}

function useConveyer(props) {
  if (props === void 0) {
    props = {};
  }

  var elementRef = useRef();
  var result = useReactive(__assign({
    data: function () {
      return {
        container: elementRef,
        props: props
      };
    }
  }, REACTIVE_CONVEYER));
  return Object.assign(elementRef, result, {
    ref: elementRef
  });
}

function useInlineConveyer(props) {
  if (props === void 0) {
    props = {};
  }

  var elementRef = useRef();
  useReactive(__assign({
    data: function () {
      return {
        container: elementRef,
        props: props
      };
    }
  }, REACTIVE_CONVEYER));
  return elementRef;
}

export { useInlineConveyer as conveyer, useConveyer };
